proxy/src/schema/registry.rs

```diff
@@ -1,82 +1,149 @@
+    /// the API consumer to query state changes of a transaction.
+    pub id: juniper::ID,
+    /// List of operations to be applied to the Registry. Currently limited to exactly one. We use
+    /// a Vec here to accommodate future extensibility.
+    pub messages: Vec<Message>,
+    /// Current state of the transaction.
+    pub state: TransactionState,
+    /// Creation time.
+    pub timestamp: String,
+}
+
+/// Required information to issue a new project registration on the [`Regisry`].
+#[derive(juniper::GraphQLObject)]
+pub struct ProjectRegistration {
+    /// The domain/namespace the project should be registered for.
+    pub domain: String,
+    /// The name of the project, which MUST be uniqure for the domain.
+    pub name: String,
+}
+
+/// Possible messages a [`Transaction`] can carry.
+pub enum Message {
+    /// Issue a new project registration. See [`ProjectRegistration`] for the shape of the data.
+    ProjectRegistration(ProjectRegistration),
+}
+
+juniper::graphql_union!(Message: () where Scalar = <S> |&self| {
+    instance_resolvers: |_| {
+        &ProjectRegistration => match *self { Message::ProjectRegistration(ref p) => Some(p) },
+    }
+});
+
+/// Possible states a [`Transaction`] can have. Useful to reason about the lifecycle and
+/// whereabouts of a given [`Transaction`].
+pub enum TransactionState {
+    /// [`Transaction`] has been applied to a block. See [`Applied`] for the shape of the data.
+    Applied(Applied),
+}
+
+juniper::graphql_union!(TransactionState: () where Scalar = <S> |&self| {
+    instance_resolvers: |_| {
+        &Applied => match *self { TransactionState::Applied(ref a) => Some(a) },
+    }
+});
+
+/// Signals that [`Transaction`] has been successfully applied by a Node to a block. Carries the
+/// hash of the Block for further inspection and/or tracking of propagation.
+#[derive(GraphQLObject)]
+pub struct Applied {
+    /// Block hash the [`Transaction`] has been applied to.
+    pub block: juniper::ID,
+}
+
 /// Registry client wrapper.
 #[derive(Clone)]
-pub struct Registry<R>
-where
-    R: ClientT,
-{
+pub struct Registry {
     /// Registry client, whether an emulator or otherwise.
-    client: R,
+    client: Client,
 }
 
 /// Registry client wrapper methods
-impl<R> Registry<R>
-where
-    R: ClientT,
-{
-    // TODO(xla): Remvoe once integrated in the schema.
-    #[allow(dead_code)]
+impl Registry {
     /// Wrap a registry client.
-    pub fn new(client: R) -> Self {
+    pub const fn new(client: Client) -> Self {
         Self { client }
     }
 
-    // TODO(xla): Remvoe once integrated in the schema.
-    #[allow(dead_code)]
     /// Register a new project on the chain.
-    pub fn register_project(
+    pub async fn register_project(
         &self,
-        author: &Pair,
-        name: String,
+        author: &ed25519::Pair,
         domain: String,
-    ) -> Result<(), Error> {
+        name: String,
+    ) -> Result<H256, Error> {
```

Comment:
You can use `radicle_registry_client::TxHash` to make the type more descriptive. (It’s an alias)